vignettes/new-guides/preprocessing_layers.Rmd

```diff
@@ -0,0 +1,636 @@
+cat("Model output: "); print(test_output); cat("\n")
+```
+
+
+
+### Encoding text as a dense matrix of ngrams with TF-IDF weighting
+
+This is an alternative way of preprocessing text before passing it to a `layer_dense` layer.
+
+```{r}
+# Define some text data to adapt the layer
+adapt_data <- as_tensor(c(
+        "The Brain is wider than the Sky",
+        "For put them side by side",
+        "The one the other will contain",
+        "With ease and You beside"
+))
+
+# Instantiate layer_text_vectorization() with "tf-idf" output_mode
+# (multi-hot with TF-IDF weighting) and ngrams=2 (index all bigrams)
+text_vectorizer = layer_text_vectorization(output_mode="tf-idf", ngrams=2)
+# Index the bigrams and learn the TF-IDF weights via `adapt()`
+text_vectorizer %>% adapt(adapt_data)
+
+# Try out the layer
+cat(
+    "Encoded text:\n", as.array(text_vectorizer("The Brain is deeper than the sea"))
+)
+
+# Create a simple model
+input <- layer_input(shape = text_vectorizer$vocabulary_size(), dtype="int64")
+output <- input %>% layer_dense(1)
+model <- keras_model(input, output)
+
+# Create a labeled dataset (which includes unknown tokens)
+train_dataset = tensor_slices_dataset(list(
+    c("The Brain is deeper than the sea", "for if they are held Blue to Blue"), 
+    c(1L, 0L)
+))
+
+# Preprocess the string inputs, turning them into int sequences
+train_dataset <- train_dataset %>% 
+  dataset_batch(2) %>% 
+  dataset_map(~list(text_vectorizer(.x), .y))
+
+
+# Train the model on the int sequences
+cat("Training model...")
+model %>% 
+  compile(optimizer="rmsprop", loss="mse") %>% 
+  fit(train_dataset)
+
+# For inference, you can export a model that accepts strings as input
+input <- layer_input(shape = 1, dtype="string")
+
+output <- input %>% 
+  text_vectorizer() %>% 
+  model()
+
+end_to_end_model = keras_model(input, output)
+
+# Call the end-to-end model on test data (which includes unknown tokens)
+cat("Calling end-to-end model on test string...\n")
+test_data <- tf$constant(matrix("The one the other will absorb"))
+test_output <- end_to_end_model(test_data)
+cat("Model output: "); print(test_output)
+```
+
+
+## Important gotchas
+
+### Working with lookup layers with very large vocabularies
+
+You may find yourself working with a very large vocabulary in a
+`layer_text_vectorization()`, a `layer_string_lookup()` layer, or an
+`layer_integer_lookup()` layer. Typically, a vocabulary larger than 500MB would
+be considered "very large".
+
+In such case, for best performance, you should avoid using `adapt()`.
+Instead, pre-compute your vocabulary in advance
+(you could use Apache Beam or TF Transform for this)
+and store it in a file. Then load the vocabulary into the layer at construction
+time by passing the filepath as the `vocabulary` argument.
+
+
+### Using lookup layers on a TPU pod or with `ParameterServerStrategy`.
+
+There is an outstanding issue that causes performance to degrade when using a
```

Comment:
This mention of TPUs makes me wonder ... have we ever tested R tensorflow on a TPU? 
Not sure, but we may want to remove all mention of TPUs from this vignette (the one above as well) ...?